pkg/kubelet/cm/device_plugin_handler.go

```diff
@@ -0,0 +1,240 @@
+}
+
+type updateCapacityCallback func(v1.ResourceList)
+
+type DevicePluginHandlerImpl struct {
+	sync.Mutex
+	devicePluginManager deviceplugin.Manager
+	// devicePluginManagerMonitorCallback is used for testing only.
+	devicePluginManagerMonitorCallback deviceplugin.MonitorCallback
+	// allDevices contains all of registered resourceNames and their exported device IDs.
+	allDevices         map[string]sets.String
+	// allocatedDevices contains pod to allocated device mapping, keyed by resourceName.
+	allocatedDevices   map[string]*podDevices
+}
+
+// NewDevicePluginHandler create a DevicePluginHandler
+// updateCapacityFunc is called to update ContainerManager capacity when
+// device capacity changes.
+func NewDevicePluginHandlerImpl(updateCapacityFunc func(v1.ResourceList)) (*DevicePluginHandlerImpl, error) {
+	glog.V(2).Infof("Creating Device Plugin Handler")
+	handler := &DevicePluginHandlerImpl{
+		allDevices:         make(map[string]sets.String),
+		allocatedDevices:   devicesInUse(),
+	}
+
+	deviceManagerMonitorCallback := func(resourceName string, added, updated, deleted []*pluginapi.Device) {
+		var capacity = v1.ResourceList{}
+		kept := append(updated, added...)
+		if _, ok := handler.allDevices[resourceName]; !ok {
+			handler.allDevices[resourceName] = sets.NewString()
+		}
+		// For now, DevicePluginHandler only keeps track of healthy devices.
+		// We can revisit this later when the need comes to track unhealthy devices here.
+		for _, dev := range kept {
+			if dev.Health == pluginapi.Healthy {
+				handler.allDevices[resourceName].Insert(dev.ID)
+			} else {
+				handler.allDevices[resourceName].Delete(dev.ID)
+			}
+		}
+		for _, dev := range deleted {
+			handler.allDevices[resourceName].Delete(dev.ID)
+		}
+		capacity[v1.ResourceName(resourceName)] = *resource.NewQuantity(int64(handler.allDevices[resourceName].Len()), resource.DecimalSI)
+		updateCapacityFunc(capacity)
+	}
+
+	mgr, err := deviceplugin.NewManagerImpl(pluginapi.KubeletSocket, deviceManagerMonitorCallback)
+	if err != nil {
+		return nil, fmt.Errorf("Failed to initialize device plugin manager: %+v", err)
+	}
+
+	handler.devicePluginManager = mgr
+	handler.devicePluginManagerMonitorCallback = deviceManagerMonitorCallback
+	return handler, nil
+}
+
+func (h *DevicePluginHandlerImpl) Start() error {
+	return h.devicePluginManager.Start()
+}
+
+func (h *DevicePluginHandlerImpl) Devices() map[string][]*pluginapi.Device {
+	return h.devicePluginManager.Devices()
+}
+
+func (h *DevicePluginHandlerImpl) Allocate(pod *v1.Pod, container *v1.Container, activePods []*v1.Pod) ([]*pluginapi.AllocateResponse, error) {
+	var ret []*pluginapi.AllocateResponse
+	h.updateAllocatedDevices(activePods)
+	for k, v := range container.Resources.Limits {
+		resource := string(k)
+		needed := int(v.Value())
+		glog.V(3).Infof("needs %d %s", needed, resource)
+		if !deviceplugin.IsDeviceName(k) || needed == 0 {
+			continue
+		}
```

Comment:
Is it needed to see if it's an existing resource? Or just rely the judgement of resource number as below?

Reply:
From performance point of view, I think it is better to skip unnecessary handling early on.